src/bgp.rs

```diff
@@ -0,0 +1,1049 @@
+pub const BGP_ROUTE_REFRESH_MSG_TYPE: u8 = 5;
+
+/// Fixed part of a BGP OPEN message (RFC 4271, Section 4.2).
+#[repr(C, packed)]
+#[derive(Debug, Copy, Clone, Default)]
+#[cfg_attr(feature = "serde", derive(::serde::Serialize, ::serde::Deserialize))]
+pub struct OpenMsgLayout {
+    /// Version: Protocol version number.
+    pub version: u8,
+    /// My Autonomous System: AS number of the sender.
+    pub my_as: [u8; 2],
+    /// Hold Time: Proposed seconds for the Hold Timer.
+    pub hold_time: [u8; 2],
+    /// BGP Identifier: IP address of the sender.
+    pub bgp_id: [u8; 4],
+    /// Optional Parameters Length: Total length of Optional Parameters.
+    pub opt_parm_len: u8,
+}
+
+impl OpenMsgLayout {
+    /// Length of the fixed part of an OPEN message in bytes.
+    pub const LEN: usize = 10;
+    const VERSION_OFFSET: usize = 0;
+    const MY_AS_OFFSET: usize = 1;
+    const HOLD_TIME_OFFSET: usize = 3;
+    const BGP_ID_OFFSET: usize = 5;
+    const OPT_PARM_LEN_OFFSET: usize = 9;
+}
+
+/// Initially fixed part of a BGP UPDATE message (RFC 4271, Section 4.3).
+#[repr(C, packed)]
+#[derive(Debug, Copy, Clone, Default)]
+#[cfg_attr(feature = "serde", derive(::serde::Serialize, ::serde::Deserialize))]
+pub struct UpdateInitialMsgLayout {
+    /// Withdrawn Routes Length: Total length of Withdrawn Routes.
+    pub withdrawn_routes_len: [u8; 2],
+}
+
+impl UpdateInitialMsgLayout {
+    /// Length of the initially fixed part of an UPDATE message in bytes.
+    pub const LEN: usize = 2;
+    const WITHDRAWN_ROUTES_LEN_OFFSET: usize = 0;
+}
+
+/// Fixed part of a BGP NOTIFICATION message (RFC 4271, Section 4.5).
+#[repr(C, packed)]
+#[derive(Debug, Copy, Clone, Default)]
+#[cfg_attr(feature = "serde", derive(::serde::Serialize, ::serde::Deserialize))]
+pub struct NotificationMsgLayout {
+    /// Error Code: Type of error.
+    pub error_code: u8,
+    /// Error Subcode: Specific information about the error.
+    pub error_subcode: u8,
+}
+
+impl NotificationMsgLayout {
+    /// Length of the fixed part of a NOTIFICATION message in bytes.
+    pub const LEN: usize = 2;
+    const ERROR_CODE_OFFSET: usize = 0;
+    const ERROR_SUBCODE_OFFSET: usize = 1;
+}
+
+/// Fixed part of a BGP ROUTE-REFRESH message (RFC 2918, Section 3).
+#[repr(C, packed)]
+#[derive(Debug, Copy, Clone, Default)]
+#[cfg_attr(feature = "serde", derive(::serde::Serialize, ::serde::Deserialize))]
+pub struct RouteRefreshMsgLayout {
+    /// Address Family Identifier (AFI)
+    pub afi: [u8; 2],
+    /// Reserved: Set to 0.
+    pub _reserved: u8,
+    /// Subsequent Address Family Identifier (SAFI).
+    pub safi: u8,
+}
+
+impl RouteRefreshMsgLayout {
+    /// Length of the fixed part of a ROUTE-REFRESH message in bytes.
+    pub const LEN: usize = 4;
+    const AFI_OFFSET: usize = 0;
+    const RES_OFFSET: usize = 2;
+    const SAFI_OFFSET: usize = 3;
+}
+
+/// BGP header and initially fixed part of its payload.
+#[repr(C, packed)]
+#[derive(Copy, Clone)]
```

Comment:
Can we add a debug trait?
```suggestion
#[derive(Debug, Copy, Clone)]
```

Reply:
I accepted your suggestion, but there is a `Debug` trait implemented for `BgpHdr`. It's lower down in the file. I found the custom one to be more human readable and easier to work with while I developed `bgp.rs`.